idx/memory/memory.go

```diff
@@ -0,0 +1,475 @@
+package memory
+
+import (
+	"flag"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/raintank/met"
+	"github.com/raintank/metrictank/idx"
+	"github.com/raintank/worldping-api/pkg/log"
+	"github.com/rakyll/globalconf"
+	"gopkg.in/raintank/schema.v1"
+)
+
+var (
+	idxOk             met.Count
+	idxFail           met.Count
+	idxAddDuration    met.Timer
+	idxGetDuration    met.Timer
+	idxListDuration   met.Timer
+	idxFindDuration   met.Timer
+	idxDeleteDuration met.Timer
+
+	Enabled bool
+)
+
+func ConfigSetup() {
+	memoryIdx := flag.NewFlagSet("memory-idx", flag.ExitOnError)
+	memoryIdx.BoolVar(&Enabled, "enabled", false, "")
+	globalconf.Register("memory-idx", memoryIdx)
+}
+
+type Tree struct {
+	Items map[string]*Node
+}
+
+type Node struct {
+	Path     string
+	Children []string
+	Leaf     bool
+}
+
+func (n *Node) String() string {
+	t := "branch"
+	if n.Leaf {
+		t = "leaf"
+	}
+	return fmt.Sprintf("%s - %s", t, n.Path)
+}
+
+// Implements the the "MetricIndex" interface
+type MemoryIdx struct {
+	sync.Mutex
+	DefById map[string]*schema.MetricDefinition
+	Tree    map[int]*Tree
```

Comment:
I was experimenting.  Originally i thought it would be fine as the metricDefs dont get updated.  But we do set the "LastUpdate" field every time a metric is seen.  see https://github.com/raintank/metrictank/pull/288/commits/ec79d6522ae84c7f7b75fa7f69e80b363abd416b